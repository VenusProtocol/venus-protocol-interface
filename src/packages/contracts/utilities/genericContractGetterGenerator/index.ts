import type { Provider } from '@ethersproject/abstract-provider';
import { Contract, ContractInterface, Signer } from 'ethers';

export interface GenericContractGetterGeneratorInput {
  abi: ContractInterface;
}

export interface GenericContractGetterInput {
  address: string;
  signerOrProvider: Signer | Provider;
}

export type GenericContractGetter<TContract extends Contract> = (
  input: GenericContractGetterInput,
) => TContract;

export const genericContractGetterGenerator = <TContract extends Contract>({
  abi,
}: GenericContractGetterGeneratorInput) => {
  const getter: GenericContractGetter<TContract> = ({ signerOrProvider, address }) =>
    new Contract(address, abi, signerOrProvider) as TContract;

  return getter;
};
