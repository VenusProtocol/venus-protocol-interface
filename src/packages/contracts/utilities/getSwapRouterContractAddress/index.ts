import addresses from 'packages/contracts/generated/infos/addresses';
import { ChainId } from 'types';

export type GetSwapRouterContractAddressInput = {
  comptrollerContractAddress: string;
  chainId: ChainId;
};

export const getSwapRouterContractAddress = ({
  comptrollerContractAddress,
  chainId,
}: GetSwapRouterContractAddressInput) => {
  const swapRouterContractAddresses = addresses.SwapRouter[chainId];
  const sanitizedComptrollerAddress =
    comptrollerContractAddress.toLowerCase() as keyof typeof swapRouterContractAddresses;

  return swapRouterContractAddresses?.[sanitizedComptrollerAddress] as string | undefined;
};
