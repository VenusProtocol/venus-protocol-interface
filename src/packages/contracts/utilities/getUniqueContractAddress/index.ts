import addresses from 'packages/contracts/generated/infos/addresses';
import { UniqueContractName } from 'packages/contracts/generated/infos/types';
import { ChainId } from 'types';

export type GetUniqueContractAddressInput = {
  name: UniqueContractName;
  chainId: ChainId;
};

export const getUniqueContractAddress = ({ name, chainId }: GetUniqueContractAddressInput) => {
  const contractAddresses = addresses[name];

  return Object.prototype.hasOwnProperty.call(contractAddresses, chainId)
    ? contractAddresses[chainId as keyof typeof contractAddresses]
    : undefined;
};
