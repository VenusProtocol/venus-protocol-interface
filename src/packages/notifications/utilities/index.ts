import { store } from '../store';
import {
  AddNotificationInput,
  Notification,
  RemoveNotificationInput,
  UpdateNotificationInput,
} from '../store/types';

export const MAX_NOTIFICATIONS = 5;
const DISPLAY_TIME_MS = 9000;

const timeoutsMapping: {
  [notificationId: Notification['id']]: NodeJS.Timer;
} = {};

const setHideTimeout = ({ id }: { id: Notification['id'] }) => {
  const { removeNotification } = store.getState();

  // Automatically hide notification after a certain time
  timeoutsMapping[id] = setTimeout(() => removeNotification({ id }), DISPLAY_TIME_MS);
};

export const hideNotification = ({ id }: RemoveNotificationInput) => {
  // Clear hide timeout if one was set
  clearTimeout(timeoutsMapping[id]);

  const { removeNotification } = store.getState();
  removeNotification({ id });
};

type DisplayNotificationInput = AddNotificationInput & {
  autoClose?: boolean;
};

export const displayNotification = ({
  autoClose = true,
  ...addNotificationInput
}: DisplayNotificationInput) => {
  const { addNotification, notifications } = store.getState();

  // Remove last notification if we've reached the maximum allowed
  if (notifications.length >= MAX_NOTIFICATIONS) {
    hideNotification({ id: notifications[notifications.length - 1].id });
  }

  // Add notification to store
  const newNotificationId = addNotification(addNotificationInput);

  if (autoClose) {
    setHideTimeout({ id: newNotificationId });
  }

  return newNotificationId;
};

type UpdateNotificationUtilInput = UpdateNotificationInput & {
  autoClose?: boolean;
};

export const updateNotification = ({
  autoClose = true,
  ...updateNotificationInput
}: UpdateNotificationUtilInput) => {
  // Clear hide timeout if one was set
  clearTimeout(timeoutsMapping[updateNotificationInput.id]);

  const { updateNotification: updateStoreNotification } = store.getState();

  // Update notification
  updateStoreNotification(updateNotificationInput);

  if (autoClose) {
    setHideTimeout({ id: updateNotificationInput.id });
  }
};
