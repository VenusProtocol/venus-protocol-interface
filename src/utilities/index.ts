export { promisify } from 'utilities/promisify';
export { restService } from 'utilities/restService';
