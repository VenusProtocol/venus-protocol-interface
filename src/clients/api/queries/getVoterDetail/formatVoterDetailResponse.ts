import BigNumber from 'bignumber.js';
import { NULL_ADDRESS } from 'constants/address';
import { getSupportName } from 'utilities';
import { IVoterDetail, VoteDetailTransactionVote, VoteDetailTransactionTransfer } from 'types';
import { IGetVoterDetailResponse } from './types';

const formatVoterResponse = (
  { balance, delegateCount, delegates, txs, votes }: IGetVoterDetailResponse,
  address: string,
): IVoterDetail => ({
  balanceWei: new BigNumber(balance),
  delegateCount,
  delegates,
  delegating: delegates !== NULL_ADDRESS && delegates !== address.toLowerCase(),
  votesWei: new BigNumber(votes),
  voterTransactions: txs.map(
    ({
      blockNumber,
      blockTimestamp,
      createdAt,
      from,
      to,
      transactionHash,
      transactionIndex,
      type,
      updatedAt,
      ...rest
    }) => {
      if (type === 'vote' && 'support' in rest) {
        return {
          blockNumber,
          blockTimestamp: new Date(blockTimestamp * 1000),
          createdAt: new Date(createdAt),
          from,
          to,
          transactionHash,
          transactionIndex,
          type,
          support: getSupportName(rest.support),
          updatedAt: new Date(updatedAt),
          votesWei: new BigNumber(votes),
        } as VoteDetailTransactionVote;
      }
      return {
        amountWei: new BigNumber(rest.amount),
        blockNumber,
        blockTimestamp: new Date(blockTimestamp * 1000),
        createdAt: new Date(createdAt),
        from,
        to,
        transactionHash,
        transactionIndex,
        type,
        updatedAt: new Date(updatedAt),
        votesWei: new BigNumber(votes),
      } as VoteDetailTransactionTransfer;
    },
  ),
});

export default formatVoterResponse;
