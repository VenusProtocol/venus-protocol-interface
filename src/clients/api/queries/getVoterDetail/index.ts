import { IVoterDetail } from 'types';
import { restService } from 'utilities';
import formatVoterDetailResponse from './formatVoterDetailResponse';
import { IGetVoterDetailResponse } from './types';

export interface IGetVoterDetailInput {
  address: string;
}

export type GetVoterDetailOutput = IVoterDetail;

const getVoterDetail = async ({ address }: IGetVoterDetailInput): Promise<GetVoterDetailOutput> => {
  const response = await restService<IGetVoterDetailResponse>({
    endpoint: `/voters/accounts/${address}`,
    method: 'GET',
  });
  const payload = response.data?.data;
  if ('result' in response && response.result === 'error') {
    // @todo Add specific api error handling
    throw new Error(response.message);
  }
  if (!payload) {
    // @todo Add specific api error handling
    throw new Error('Unexpected error retrieving voter details');
  }

  return formatVoterDetailResponse(payload, address);
};

export default getVoterDetail;
