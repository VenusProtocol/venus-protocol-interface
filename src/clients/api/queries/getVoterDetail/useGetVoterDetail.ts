import { useQuery, QueryObserverOptions } from 'react-query';
import getVoterDetail, {
  IGetVoterDetailInput,
  GetVoterDetailOutput,
} from 'clients/api/queries/getVoterDetail';
import FunctionKey from 'constants/functionKey';

type Options = QueryObserverOptions<
  GetVoterDetailOutput,
  Error,
  GetVoterDetailOutput,
  GetVoterDetailOutput,
  [FunctionKey.GET_VOTER_DETAIL, IGetVoterDetailInput]
>;

const useGetVoterDetail = (params: IGetVoterDetailInput, options?: Options) =>
  // This endpoint is paginated so we keep the previous responses by default to create a more seamless paginating experience
  useQuery([FunctionKey.GET_VOTER_DETAIL, params], () => getVoterDetail(params), options);

export default useGetVoterDetail;
