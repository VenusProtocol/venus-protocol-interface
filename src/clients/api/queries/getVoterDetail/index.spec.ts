import { restService } from 'utilities';
import voterDetailResponse from '__mocks__/api/voterDetail.json';
import fakeAddress from '__mocks__/models/address';
import getVoterDetail from '.';

jest.mock('utilities/restService');

describe('api/queries/getVoterDetail', () => {
  test('throws an error when request fails', async () => {
    const fakeErrorMessage = 'Fake error message';

    (restService as jest.Mock).mockImplementationOnce(async () => ({
      result: 'error',
      status: false,
      message: fakeErrorMessage,
    }));

    try {
      await getVoterDetail({ address: fakeAddress });

      throw new Error('getVoterDetail should have thrown an error but did not');
    } catch (error) {
      expect(error).toMatchInlineSnapshot('[Error: Fake error message]');
    }
  });

  test('returns  formatted voter details', async () => {
    (restService as jest.Mock).mockImplementationOnce(async () => ({
      status: 200,
      data: voterDetailResponse,
    }));

    const response = await getVoterDetail({ address: fakeAddress });

    expect(restService).toBeCalledWith({
      endpoint: `/voters/accounts/${fakeAddress}`,
      method: 'GET',
    });

    expect(response).toMatchSnapshot();
  });
});
