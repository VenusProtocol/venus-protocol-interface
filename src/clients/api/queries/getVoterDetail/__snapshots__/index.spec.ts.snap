// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/queries/getVoterDetail returns  formatted voter details 1`] = `
Object {
  "balance": "0",
  "delegateCount": 1,
  "delegates": "0x315deBE419d9f9aec19807609b396Ef33a9c44C0",
  "transactions": Array [
    Object {
      "amountWei": "2470334759651375154",
      "blockNumber": 18051658,
      "blockTimestamp": 1653308122,
      "createdAt": 2022-05-23T12:15:31.000Z,
      "from": "0x315debe419d9f9aec19807609b396ef33a9c44c0",
      "to": "0x7eb5d86fd78f3852a3e0e064f2842d45a3db6ea2",
      "transactionHash": "0x1d6c5d29ba91dbe1e6150d2335bba9a2b99bf850b1c5255ed76e076bc8f0932c",
      "transactionIndex": 19,
      "type": "transfer",
      "updatedAt": 2022-05-23T12:15:31.000Z,
    },
    Object {
      "amountWei": "2470307000576804910",
      "blockNumber": 18051572,
      "blockTimestamp": 1653307864,
      "createdAt": 2022-05-23T12:11:10.000Z,
      "from": "0xfd36e2c2a6789db23113685031d7f16329158384",
      "to": "0x315debe419d9f9aec19807609b396ef33a9c44c0",
      "transactionHash": "0xb312250c6f5d4c288d79c9855dc7fceed8c5f10b4b8b1fec66130deadafa21ab",
      "transactionIndex": 201,
      "type": "transfer",
      "updatedAt": 2022-05-23T12:11:30.000Z,
    },
    Object {
      "amountWei": "3019084454051545384",
      "blockNumber": 17824116,
      "blockTimestamp": 1652619724,
      "createdAt": 2022-05-15T13:02:30.000Z,
      "from": "0x315debe419d9f9aec19807609b396ef33a9c44c0",
      "to": "0x7eb5d86fd78f3852a3e0e064f2842d45a3db6ea2",
      "transactionHash": "0x860625c320493e7f900017e4ff6d2700c5def8b81aaef16e0adcbcd2f9100b9a",
      "transactionIndex": 85,
      "type": "transfer",
      "updatedAt": 2022-05-15T13:02:30.000Z,
    },
    Object {
      "amountWei": "85354896669975801",
      "blockNumber": 17824104,
      "blockTimestamp": 1652619685,
      "createdAt": 2022-05-15T13:01:40.000Z,
      "from": "0xfd36e2c2a6789db23113685031d7f16329158384",
      "to": "0x315debe419d9f9aec19807609b396ef33a9c44c0",
      "transactionHash": "0x2dffbd3f586f3ecf7b6309a9a566c1a9f5da7cec97e6e466172ccb2d59452a70",
      "transactionIndex": 131,
      "type": "transfer",
      "updatedAt": 2022-05-15T13:01:41.000Z,
    },
  ],
  "votesWei": "0",
}
`;
